crates/mysten-network/src/client.rs

```diff
@@ -107,20 +125,51 @@ impl MyEndpoint {
         });
 
         if disable_caching_resolver {
-            self.endpoint.connect_lazy()
+            if let Some(tls_config) = self.tls_config {
+                self.endpoint.connect_with_connector_lazy(
+                    hyper_rustls::HttpsConnectorBuilder::new()
+                        .with_tls_config(tls_config)
+                        .https_only()
+                        .enable_http2()
+                        .build(),
+                )
+            } else {
+                self.endpoint.connect_lazy()
+            }
         } else {
             let mut http = HttpConnector::new_with_resolver(CachingResolver::new());
             http.enforce_http(false);
             http.set_nodelay(true);
             http.set_keepalive(None);
             http.set_connect_timeout(None);
 
-            self.endpoint.connect_with_connector_lazy(http)
+            if let Some(tls_config) = self.tls_config {
+                let https = hyper_rustls::HttpsConnectorBuilder::new()
+                    .with_tls_config(tls_config)
+                    .https_only()
+                    .enable_http1()
+                    .wrap_connector(http);
+                self.endpoint.connect_with_connector_lazy(https)
+            } else {
+                self.endpoint.connect_with_connector_lazy(http)
+            }
         }
     }
 
     async fn connect(self) -> Result<Channel> {
-        self.endpoint.connect().await.map_err(Into::into)
+        if let Some(tls_config) = self.tls_config {
+            let https_connector = hyper_rustls::HttpsConnectorBuilder::new()
+                .with_tls_config(tls_config)
+                .https_only()
+                .enable_http2()
+                .build();
+            self.endpoint
+                .connect_with_connector(https_connector)
+                .await
+                .map_err(Into::into)
```

Comment:
Instead of building the https connector ourselves, lets use the same thing that consensus uses,`tonic_rustls`. Then we can unconditionally build the channel with the expected cert but instead continue using 'http' urls instead of 'https' urls. Then when we're ready to start using this in prod all we have to do it flip the schema on the url we use.

Reply:
I don't think we want TLS or not to be selected based on URL schema. Otherwise any existing clients that erroneously have "https" schemas now will stop working once this change goes in until validators have upgraded. I'd rather force it to stay on TCP for now, and then force it over to TLS once the server-side support rolls out. (Then most likely remove server-side support for unencrypted TCP.)

Reply:
What I'm trying to highlight is that eventually we want the url used to be https when we are using tls and http when we aren't.

I suppose the logistics of how we do that is less important. I just want this to be clear if those urls show up in logs

Reply:
What I'm saying is I think we don't want that. We should completely ignore the URL schema (which people probably set arbitrarily without even thinking about it) and do what we think is right, which is stick with TCP now and then force TLS once it's available.

Reply:
Right now these URLs are coming from the on chain system state, which would mean that if we respect the URL schema, individual validators would be the ones controlling whether clients connect to them via TCP or TLS. That's def not what we want.

Reply:
I'm fine if we "force" connections to be TLS when its available, but the URL schema conveys intent and we should, in code we control, ensure we are properly conveying that intent where possible. This means when we do move to TLS, we should make sure that these URLs we construct to connect with the validators should have 'https' schemas and not 'http' ones.

Reply:
Sure, it would involve overwriting the addresses pulled from the system state and given this PR is already fairly large maybe we can save that for a follow-up?

Reply:
(I'll note that this makes things no worse than they already are, as we currently will ignore "https" schemas and still connect via "http")